sys/exec.c

```diff
@@ -0,0 +1,292 @@
+              (void *)ph->p_vaddr, (unsigned int)ph->p_offset,
+              (unsigned int)ph->p_filesz, (unsigned int)ph->p_memsz,
+              (unsigned int)ph->p_flags);
+      if (ph->p_vaddr % PAGESIZE) {
+        kprintf("[exec] Exec failed: Segment p_vaddr is not page alligned\n");
+        goto exec_fail;
+      }
+      vm_addr_t start = ph->p_vaddr;
+      vm_addr_t end = roundup(ph->p_vaddr + ph->p_memsz, PAGESIZE);
+      // TODO: What if segments overlap?
+      // Temporarily permissive protection.
+      vm_map_entry_t *segment = vm_map_add_entry(
+        vmap, start, end, VM_PROT_READ | VM_PROT_WRITE | VM_PROT_EXEC);
+      // Allocate pages backing this segment.
+      segment->object = default_pager->pgr_alloc();
+      // Copy data into the segment
+      memcpy((uint8_t *)start, elf_image + ph->p_offset, ph->p_filesz);
+      // Zero the rest
+      if (ph->p_filesz < ph->p_memsz) {
+        bzero((uint8_t *)start + ph->p_filesz, ph->p_memsz - ph->p_filesz);
+      }
+      // Apply correct permissions
+      vm_prot_t prot = VM_PROT_NONE;
+      if (ph->p_flags | PF_R)
+        prot |= VM_PROT_READ;
+      if (ph->p_flags | PF_W)
+        prot |= VM_PROT_WRITE;
+      if (ph->p_flags | PF_X)
+        prot |= VM_PROT_EXEC;
+      // Note: vm_map_protect is not yet implemented, so
+      // this will have no effect as of now
+      vm_map_protect(vmap, start, end, prot);
+    }
+  }
+
+  // Create a stack segment. As for now, the stack size is fixed and
+  // will not grow on-demand. Also, the stack info should be saved
+  // into the thread structure.
+  // Generally, the stack should begin at a high address (0x80000000),
+  // excluding env vars and arguments, but I've temporarly moved it
+  // a bit lower so that it is easier to spot invalid memory access
+  // when the stack underflows.
+  vm_addr_t stack_bottom = 0x70000000;
+  const size_t stack_size = PAGESIZE * 2;
+
+  vm_addr_t stack_start = stack_bottom - stack_size;
+  vm_addr_t stack_end = stack_bottom;
+  // TODO: What if this area overlaps with a loaded segment?
+  vm_map_entry_t *stack_segment = vm_map_add_entry(
+    vmap, stack_start, stack_end, VM_PROT_READ | VM_PROT_WRITE);
+  stack_segment->object = default_pager->pgr_alloc();
+
+  // Prepare program stack, which includes storing program args.
+  prepare_program_stack(args, &stack_bottom);
+  kprintf("[exec] Stack real bottom at %p\n", (void *)stack_bottom);
+
+  vm_map_dump(vmap);
+
+  // At this point we are certain that exec suceeds.
+  // We can safely destroy the previous vm map.
+
+  // This condition will be unnecessary when we take the vmap info
+  // from thread struct. The user thread calling exec() WILL
+  // certainly have an existing vmap before exec()ing.
+  if (old_vmap)
+    vm_map_delete(old_vmap);
+
+  // TODO: Assign the new vm map to the process structure
+
+  // We need to correct the thread structure so that it can hold the
+  // correct $gp value for this execution context.
```

Comment:
I don't think we have to care about it... please read about *gp in non-PIC code* from *MIPS ABIs Described*, page 7.